Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift

```diff
@@ -0,0 +1,750 @@
+        @usableFromInline
+        /* private */ internal var storage: Storage
+
+        @inlinable
+        init(storage: Storage) {
+            self.storage = storage
+        }
+
+        /// Toggles the writability of the ``NIOAsyncWriter``.
+        ///
+        /// If the writer becomes writable again all suspended yields will be resumed and the produced elements will be forwarded via
+        /// the ``NIOAsyncWriterDelegate/didYield(contentsOf:)`` method.
+        public func toggleWritability() {
+            self.storage.toggleWritability()
+        }
+    }
+}
+
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+extension NIOAsyncWriter {
+    /// This is the underlying storage of the writer. The goal of this is to synchronize the access to all state.
+    @usableFromInline
+    /* fileprivate */ internal final class Storage: @unchecked Sendable {
+        /// The lock that protects our state.
+        @usableFromInline
+        /* private */ internal let lock = Lock()
+        /// The counter used to assign an ID to all our yields.
+        @usableFromInline
+        /* private */ internal let yieldIDCounter = ManagedAtomic<UInt64>(0)
+        /// The state machine.
+        @usableFromInline
+        /* private */ internal var stateMachine: StateMachine
+        /// The delegate.
+        @usableFromInline
+        /* private */ internal var delegate: Delegate?
+
+        @usableFromInline
+        /* fileprivate */ internal init(
+            isWritable: Bool,
+            delegate: Delegate
+        ) {
+            self.stateMachine = .init(isWritable: isWritable)
+            self.delegate = delegate
+        }
+
+        @inlinable
+        /* fileprivate */ internal func writerDeinitialized() {
+            let delegate: Delegate? = self.lock.withLock {
+                let action = self.stateMachine.writerDeinitialized()
+
+                switch action {
+                case .callDidTerminate:
+                    let delegate = self.delegate
+                    self.delegate = nil
+
+                    return delegate
+
+                case .none:
+                    return nil
+                }
+            }
+
+            delegate?.didTerminate(completion: .finished)
+        }
+
+        @inlinable
+        /* fileprivate */ internal func toggleWritability() {
+            // We are manually locking here since we need to use both the delegate and the
+            // suspendedYields. Doing this with a withLock becomes very unhandy.
+            self.lock.lock()
+            let action = self.stateMachine.toggleWritability()
+
+            switch action {
+            case .callDidYieldAndResumeContinuations(let suspendedYields):
+                let delegate = self.delegate
+                self.lock.unlock()
+
+                delegate?.didYield(contentsOf: suspendedYields.lazy.map { $0.elements }.flatMap { $0 })
+                suspendedYields.forEach { $0.continuation.resume() }
```

Comment:
I prefer keeping the `forEach` since it is more concise for a very simple task. However, in general I agree to prefer `for ...in` over `forEach` for anything that is more complex